ObjC AnyPromise and SwiftPackageManager (SPM)

DG0BAB opened the pull request:
This PR makes `AnyPromise` available to Objective-C and Swift code when used with SPM.

To achieve this, the PromiseKit package now consists of two targets. PromiseKit and PromiseKitObjC. The existence of PromiseKitObjC is almost irrelevant to the client because a simple `#import PromiseKit.h` can be used in Objective-C files to get access to `AnyPromise`. If the client wants to use modules, `@import PromiseKitObjC;` must be used. 

`AnyPromise` is now defined in Swift and uses the `@objc` attribute. It's made accessible to Objective-C, by adding the category `@interface AnyPromise (ObjC)` to it. All ObjC functionality defined in this category, delegates to the Swift `AnyPromise`. 

Additional changes to other files were necessary to make everything work in the PromiseKit Xcode Project and while build as a SwiftPackage. All tests succeeded in the Xcode Project.

DG0BAB committed:
Made AnyPromise available as a "normal" type to both Swift and Objective-C even when used with Swift Packe Manager.

DG0BAB committed:
Split up the PromiseKit package into two targets. One for pure Swift and another for Objective-C
Targets must be Swift or Objective-C only. Mixing languages in one target is not supported by SPM.
Both targets get integrated into the PromiseKit library product

DG0BAB committed:
Set source path
Set public header path so that Objective-C projects can include and find "PromiseKit.h"
In Objective-C projects @import PromiseKitObjC; can be used as well.

DG0BAB committed:
Added missing import

DG0BAB committed:
Reversed some earlier changes because a forward declaration of AnyPromise is sufficient at this point.

DG0BAB committed:
Moved AnyPromise.swift to the Sources.swift group

DG0BAB committed:
Some commenting

DG0BAB committed:
Changed visibility of box to internal because it it used in an extension to AnyPromise in the file CustomStringConvertible.swift

DG0BAB committed:
Changed podspec to include my GitHub-URL

DG0BAB committed:
Reversed changes to podspec file

DG0BAB commented:
As expected, the tests and/or build on Linux failed.

DG0BAB committed:
Try to fix Travis build

DG0BAB commented:
My mistake: Not the Linux-Tests are failing. It's the Pretest and I got irritated by the `--generate-linuxmain` option....

DG0BAB committed:
Minor change to the dependencies

DG0BAB committed:
Removed obsolete code

DG0BAB committed:
Changed preprocessor conditional imports

DG0BAB committed:
Fixed typo in dependencies
Fixed formatting

DG0BAB commented:
I tested it with Xcode 10.3 and got the same results as seen here while building with Travis. I currently don't know how to fix this and I don't have the time to further investigate this issue. However it works fine with Xcode 12. So maybe this PR could be used for some a later or special SPM version of PromiseKit. I don't know. Any help for fixing this Issue is greatly appreciated.

mxcl commented:
Rebase for new CI infrastructure

mxcl commented:
This is very intrusive. I cannot merge.

DG0BAB commented:
Yeah, a lot of changes were necessary to make it work. I’m already using my fork in production code for two of my projects and it works until any problems so far. 

If you are interested in merging let me know how to help.

mxcl commented on Tests/Bridging/Infrastructure.swift:
```diff
@@ -1,3 +1,4 @@
+import Foundation
```
Presuambly unnecessary.

DG0BAB replied:
I'm sure there was a compiler error but I'll check again....

mxcl commented on Sources/fwd.h:
```diff
@@ -2,6 +2,7 @@
 #import <dispatch/dispatch.h>
 
 @class AnyPromise;
+
```
Unnecessary.

DG0BAB replied:
The newline or the forward declaration?

mxcl commented:
We need to run the full CI against this. I'll make that possible with a comment or something.

mxcl commented:
/ci

DG0BAB commented on Sources/AnyPromise.h:
```diff
@@ -2,15 +2,35 @@
+// silence a compile error which would avoid building
+// the Swift-Module on first builds.
+
+#if __has_include("PromiseKit-Swift.h")
```
The Problem here was, that AnyPromise was not always defined at this point. But it must be defined here, because otherwise I could not declare a category on it. The include path is always different depending on what is currently build. Tests, Package, Xcode Project, etc. So I figured out that this seemed to work. I'm not very satisfied with it but finally it worked. Would appreciate a better solution. 

`self->d` isn't used anymore.  The idea I was following was defining `AnyPromise` in Swift and declaring a category in ObjC on that and delegate from the category to the Swift implementation.

DG0BAB committed:
Changed package.swift for swift versions >= 5.3 to also include the PromiseKitObjC target.
This makes AnyPromise available when PromiseKit is included with SPM for Swift versions >= 5.3

DG0BAB committed:
Fixed errors while resolving includes

mrcararia commented:
Hi, 

is there any update about this issue? 
How is possibile to use AnyPromise in an ObjC class using the SPM version?

Thanks.

mxcl commented:
> is there any update about this issue?

This PR is too much. I have to maintain the results and this is beyond my ability to comprehend it without significant work. So I can't merge it.

> How is possibile to use AnyPromise in an ObjC class using the SPM version?

It is not possible without a patch like this. You could use the fork of OP. SwiftPM is decentralized by design.
